ql/src/java/org/apache/hadoop/hive/ql/metadata/MaterializedViewsCache.java

```diff
@@ -205,4 +212,52 @@ HiveRelOptMaterialization get(String dbName, String viewName) {
   public boolean isEmpty() {
     return materializedViews.isEmpty();
   }
+
+
+  private static class ASTKey {
+    private final ASTNode root;
+
+    public ASTKey(ASTNode root) {
+      this.root = root;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || getClass() != o.getClass()) return false;
+      ASTKey that = (ASTKey) o;
+      return equals(root, that.root);
+    }
+
+    private boolean equals(ASTNode astNode1, ASTNode astNode2) {
+      if (!(astNode1.getType() == astNode2.getType() &&
+              astNode1.getText().equals(astNode2.getText()) &&
+              astNode1.getChildCount() == astNode2.getChildCount())) {
+        return false;
+      }
+
+      for (int i = 0; i < astNode1.getChildCount(); ++i) {
+        if (!equals((ASTNode) astNode1.getChild(i), (ASTNode) astNode2.getChild(i))) {
+          return false;
+        }
+      }
+
+      return true;
+    }
+
+    @Override
+    public int hashCode() {
+      return hashcode(root);
+    }
+
+    private int hashcode(ASTNode node) {
+      int result = Objects.hash(node.getType(), node.getText());
```

Comment:
`ASTNode` does not contains the whole query string just small tokens extracted by the parser. If only a portion of the definition should be used for computing the hash code the limit could be the depth of the AST.
I would do it in a follow-up patch if this turn to be a bottleneck.

Reply:
Sounds ok to do in a follow-up patch if there is a performance issue.  It would be useful to leave a comment in the code about possibly revisiting this if needed.  Also,  may want to mention about the mutability aspect of ASTNode since it means that caching the hash code is not an option if something changed in the lower levels of the AST.